alg/teca_dataset_diff.cxx

```diff
@@ -168,86 +170,94 @@ const_p_teca_dataset teca_dataset_diff::execute(
+    if (!ds0->empty() && ds1->empty())
     {
         TECA_ERROR("dataset 2 is empty, 1 is not.")
         return nullptr;
     }
 
     // If the datasets are both empty, they are "equal." :-/
-    if (input_data[0]->empty() && input_data[1]->empty())
+    if (ds0->empty() && ds1->empty())
         return nullptr;
 
-    // get the inputs. They can be tables or cartesian meshes.
-    const_p_teca_table table1 =
-        std::dynamic_pointer_cast<const teca_table>(input_data[0]);
-
-    const_p_teca_table table2 =
-         std::dynamic_pointer_cast<const teca_table>(input_data[1]);
-
-    const_p_teca_cartesian_mesh mesh1 =
-        std::dynamic_pointer_cast<const teca_cartesian_mesh>(input_data[0]);
-
-    const_p_teca_cartesian_mesh mesh2 =
-        std::dynamic_pointer_cast<const teca_cartesian_mesh>(input_data[1]);
-
-    // No mixed types!
-    if (((table1 && !table2) || (!table1 && table2)) ||
-        ((mesh1 && !mesh2) || (!mesh1 && mesh2)))
+    // compare the inputs. the type of data is inferred from the
+    // reference mesh.
+    if (dynamic_cast<const teca_table*>(ds0.get()))
     {
-        TECA_ERROR("input datasets must have matching types.");
-        return nullptr;
+        if (this->compare_tables(
+            std::dynamic_pointer_cast<const teca_table>(ds0),
+            std::dynamic_pointer_cast<const teca_table>(ds1)))
+        {
+            TECA_ERROR("Failed to compare tables.");
+            return nullptr;
+        }
     }
-
-    if (!table1 && !mesh1)
+    else if (dynamic_cast<const teca_cartesian_mesh*>(ds0.get()))
     {
-        TECA_ERROR("input datasets must be teca_tables or teca_cartesian_meshes.")
-        return nullptr;
+        if (this->compare_cartesian_meshes(
+            std::dynamic_pointer_cast<const teca_cartesian_mesh>(ds0),
+            std::dynamic_pointer_cast<const teca_cartesian_mesh>(ds1)))
+        {
+            TECA_ERROR("Failed to compare cartesian_meshes.");
+            return nullptr;
+        }
     }
-
-    if (table1)
+    else if (dynamic_cast<const teca_curvilinear_mesh*>(ds0.get()))
     {
-        if (this->compare_tables(table1, table2))
+        if (this->compare_curvilinear_meshes(
+            std::dynamic_pointer_cast<const teca_curvilinear_mesh>(ds0),
+            std::dynamic_pointer_cast<const teca_curvilinear_mesh>(ds1)))
         {
-            TECA_ERROR("Failed to compare tables.");
+            TECA_ERROR("Failed to compare curvilinear_meshes.");
             return nullptr;
         }
     }
-    else
+    else if (dynamic_cast<const teca_arakawa_c_grid*>(ds0.get()))
     {
-        if (this->compare_cartesian_meshes(mesh1, mesh2))
+        if (this->compare_arakawa_c_grids(
+            std::dynamic_pointer_cast<const teca_arakawa_c_grid>(ds0),
+            std::dynamic_pointer_cast<const teca_arakawa_c_grid>(ds1)))
         {
-            TECA_ERROR("Failed to compare cartesian meshes.");
+            TECA_ERROR("Failed to compare arakawa_c_grids.");
```

Comment:
The error wording here and on lines 241, 231, 221, etc. is ambiguous: does it mean that the comparison failed, as in the comparison could not be done; or does it mean that the comparison yielded a significant difference?  I think it's the latter, in which case an alternate wording might be "Failure because arakawa_c_grids differ"

Reply:
Good point. glad to improve the error message.